design/src/rfcs/rfc0018_logging_sensitive.md

```diff
@@ -0,0 +1,407 @@
+
+These wrappers should be provided alongside the `Sensitive` struct described in [Debug Logging](#debug-logging). If they are implemented on top of `Sensitive`, they will inherit the same behavior - allowing redactions to be toggled using `unredacted-logging` feature flag.
+
+### Middleware Position
+
+This logging middleware should be applied outside of the [OperationHandler](https://github.com/awslabs/smithy-rs/blob/cd0563020abcde866a741fa123e3f2e18e1be1c9/rust-runtime/inlineable/src/server_operation_handler_trait.rs#L17-L21) after its construction in the (generated) `operation_registry.rs` file. The middleware should preserve the associated types of the `OperationHandler` (`Response = Response<BoxBody>`, `Error = Infallible`) to cause minimal disruption.
+
+An easy position to apply the logging middleware is illustrated below in the form of `Logging{Operation}::new`:
+
+```rust
+let empty_operation = LoggingEmptyOperation::new(operation(registry.empty_operation));
+let get_pokemon_species = LoggingPokemonSpecies::new(operation(registry.get_pokemon_species));
+let get_server_statistics = LoggingServerStatistics::new(operation(registry.get_server_statistics));
+let routes = vec![
+    (BoxCloneService::new(empty_operation), empty_operation_request_spec),
+    (BoxCloneService::new(get_pokemon_species), get_pokemon_species_request_spec),
+    (BoxCloneService::new(get_server_statistics), get_server_statistics_request_spec),
+];
+let router = aws_smithy_http_server::routing::Router::new_rest_json_router(routes);
+```
+
+Although an acceptable first step, putting logging middleware here is suboptimal - the `Router` allows a `tower::Layer` to be applied to the operation by using the [Router::layer](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/src/routing/mod.rs#L146) method. This middleware will be applied _outside_ of the logging middleware and, as a result, will not be subject to the span of any middleware. Therefore, the `Router` must be changed to allow for middleware to be applied within the logging middleware rather than outside of it.
+
+This is a general problem, not specific to this proposal. For example, [Use Request Extensions](#use-request-extensions) must also solve this problem.
+
+Fortunately, this problem is separable from the actual implementation of the logging middleware and we can get immediate benefit by application of it in the suboptimal position described above.
+
+### Logging within the Router
+
+There is need for logging within the `Router` implementation - this is a crucial area of business logic. As mentioned in the [Routing](#routing) section, we are permitted to log potentially sensitive data in cases where requests fail to get routed to an operation.
```

Comment:
Will we be able to correlate the log statement of a request that failed to be routed successfully (where we log potentially sensitive data) with previous log statements pertaining to the same request (where we are not allowed to log anything potentially sensitive)?

Reply:
I don't think there is a general solution we can enforce to provide this correlation. The customer is always free to apply middleware to the outside of the `Router` and log in it anyway they want.

For example:
```rust
let svc = router
    .map_request(|request| {
            if request.len() > 5 { info!("hello world"); }
            request
    });
```

What the customer could do is to attach a unique ID to a span which is open during a requests lifetime. The unique ID would adjoined to all the logging messages and so would provide the correlation. You'd then get a log like

```
span(id = 342432)
    -> info(message = "hello world")
    -> error(message = "failed to route")
```

@drganjoo was eluding to this the other day. https://github.com/awslabs/smithy-rs/pull/1536#issuecomment-1192766074 relates to it.